### PR TITLE
bpo-34659: Adds initial kwarg to itertools.accumulate()

### PR DESCRIPTION
**Pull request by lisroach**
https://bugs.python.org/issue34659

https://bugs.python.org/issue34659

### CODE REVIEW

**Review by rhettinger — APPROVED** [review]

**Review by tim-one — APPROVED** [review]
Looks good!  Thank you - I'll even use it ;-)

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
There are following technical problems with current changes:
1. The default value for *initial* is documented and exposed in the signature as `None`, but passing `None` is not the same as omitting the *initial* argument.
2. Pickling and copying the `accumulate` iterator loses the initial value.

**Comment by rhettinger** [comment]
The pickling and copying should be a separate tracker item for Kristjan Jonnson to update his own code (I think he is the only user of that feature and would be willing to shoulder some of the recurring maintenance burden arising from that code).

**Comment by serhiy-storchaka** [comment]
I don't think that ignoring `initial=None` is right. And you did not think this too, according to your [initial implementation](https://mail.python.org/pipermail/python-ideas/2018-April/049649.html).

**Comment by rhettinger** [comment]
It depends on whether you think *None* would be a useful input value to an accumulation.

The `start=_sentinel` approach gives more flexibility but is problematic for complicating the pure python version in the docs and for wrestling with argument clinic (I suspect we would have to remove the clinic code entirely and forgo the signature it generates).  If so, I'm inclined to leave the PR as is.

**Comment by serhiy-storchaka** [comment]
This will make a subtle difference between `accumulate()` and `reduce()`, and also `max()`, `iter()`, `next()`, `getattr()`, etc which accept an optional argument, but distinguish `None` from no argument. I think this will be a bug magnet.

As for the pure python version in the docs, the pure python version of `reduce()` uses `None` as a sentinel.

**Comment by rhettinger** [comment]
Added pickle support (when *initial* is set, use `chain([initial], it)`).  Please take a look.

File: Modules/itertoolsmodule.c
Comment (serhiy-storchaka) [inline Modules/itertoolsmodule.c]: `"O(NO)O"`

File: Modules/itertoolsmodule.c
Comment (serhiy-storchaka) [inline Modules/itertoolsmodule.c]: Please add spaces around `?` and `:` for PEP 7.

**Review by serhiy-storchaka — COMMENTED** [review]
Please add an entry in What's New.

File: Modules/itertoolsmodule.c
Comment (rhettinger) [reply Modules/itertoolsmodule.c]: It's better without the spaces to show the grouping and for consistency with the surrounding code.

File: Modules/itertoolsmodule.c
Comment (rhettinger) [reply Modules/itertoolsmodule.c]: IMO this is to minor to warrant a WhatsNew entry.  Whatsnew should be much more selective.  The reason we have whatsnew is that NEWS is already unmanageably lengthy to read.

Comment (rhettinger) [reply Modules/itertoolsmodule.c]: Fixed.

**Comment by serhiy-storchaka** [comment]
The pickling part LGTM. But I think that `initial=None` should not be ignored.

**Comment by rhettinger** [comment]
Serhiy, thank you for the suggestion about `initial=None`, but I'm going to decline.  I want to keep the argument clinic here and I agree with [Tim's post on the subject](https://bugs.python.org/msg325616). As module maintainer, I get to make the final call on the API.

Let's clear the way for Lisa to apply her first patch own her own.  Please clear the review request so this can move forward. It has already consumed far too much time in relative to the value of the feature.

**Comment by serhiy-storchaka** [comment]
This feature is just too complex for its value.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**
